Added g_list_free_full in lib/compat/glib & configure.ac

definit3 opened the pull request:
Fixes issue #1270  
SLES 11 has glib version <=2-26, in which g_list_free_full function is not present.
Added compat : glib patch to fix the issue.

kira-syslogng commented:
This user does not have permission to start the build. Can one of the admins verify this patch and start the build? (admin please type: ok to test)

furiel commented:
@kira-syslogng ok to test

Kokan commented:
@kira-syslogng ok to test

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

furiel commented:
@definit3 thank you for your contribution!

- When I tried your patch, it is indeed passed g_list_free_full. But syslog-ng still not compiles with glib 2.26: g_hash_table_contains is also missing. Probably g_list_free_full was only the first place that stopped the compiler. So to close #1270 we need to fix all such compile errors.

- Please edit your commits according to style check! The function declaration should be one line. One space after comma, no double spaces before function parameter. Probably our style check test is buggy, it should have detected these.

- We have both autotools and cmake. It would be great if we could maintain feature parity between autotools and cmake.
As we are talking about supporting a very old glib version, and cmake is still experimental, I don't want to force it too much. But I would be happy to see it working with cmake as well... Can you look into it?

definit3 commented:
@furiel  Okay, going to look into other errors. About the style check, is there any doc that briefly state the coding standards? ( Sorry, really new to open source) .

gaborznagy commented:
@definit3:
Generally we use a style similar to GNU coding style
https://www.gnu.org/prep/standards/standards.html#Formatting

Besides you can always check and correct your style format during
development with:
`make style-check` and `make style-format`


On Wed, Feb 28, 2018 at 11:11 AM, Vivek Raj <notifications@github.com>
wrote:

> @furiel <https://github.com/furiel> Okay, going to look into other
> errors. About the style check, is there any doc that briefly state the
> coding standards? ( Sorry, really new to open source) .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/1897#issuecomment-369191288>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYbP5XzjUOI4ewnXPIhEVfEVlmA8AACDks5tZSY7gaJpZM4SVOoH>
> .
>

gaborznagy commented:
@furiel 
Our style check test is not buggy, it's just we don't have some astyle options that would cover this.
The suitable astyle option would be `--pad-oper`, which would  `Insert space padding around operators. This will also pad commas.`
However this option can broke our style in some files, even Astyle states that some of their formatting is not [fail-safe](http://astyle.sourceforge.net/astyle.html#_Disable_Formatting)

If we only want to shot for commas, later versions of astyle (>= 2.0.6) has `--pad-comma`, but unfortunately Travis uses 2.0.5...

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

definit3 commented:
@furiel So, I tried to make a similar patch for that in compat for g_hash_table_contains but i am encountering several errors. The function g_hash_table_contains is only used once in secret_storage.c, So I replaced the g_hash_table_contains function with 'g_hash_table_lookup_extended'. Both functions does the same thing. Everything compiled successfully and I was able to run syslog-ng.  Is it okay? as i have changed a bit in secret_storage not in compat.

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

furiel commented on lib/secret-storage/secret-storage.c:
```diff
@@ -227,8 +227,9 @@ gboolean
+  gpointer *ignore_orig_key, *ignore_value;
   SecretStorage *secret_storage;
-  if (!g_hash_table_contains(secret_manager, key))
+  if (!g_hash_table_lookup_extended(secret_manager, key, ignore_orig_key, ignore_value))
```
Changing g_hash_table_contains is fine for me. Though with g_hash_table_lookup_extended there is a problem here, ignore_orig_key and ignore_value should be output parameters, meaning you should have passed &ignore_orig_key and &ignore_value. With the current version, glib would write  data to random location based on the garbage picked from the stack.

Though I would suggest something else: you can safely use g_hash_table_lookup here. From architecture point of view, either when adding a secret, either someone registers for a secret, we will allocate a buffer that is put into the hashtable. Meaning it is not possible to have NULL as value for any key (we even crash syslog-ng if the buffer could not be allocated). In short, you can use g_hash_table_lookup, there is no need for the extra safety of g_hash_table_lookup_extended.

viliampucik replied:
There is just one pitfall with RHEL6 which glib2 is missing `g_hash_table_lookup`. Any way to solve it?

furiel commented:
Now the code compiles fine for me.

But the problem is there is a unit test that also fails to compile due to missing implementation of g_mkdtemp. The test we are talking about is test_directory_monitor.c, which depends on criterion, so @definit3 if you did not install criterion-dev, you did not have a chance to catch this (advertising #1878 :) ).

Now I am not sure if we want to go that far to support old glib even for unit tests. I checked the spec file of @czanik in OBS, and I do not see criterion there either, so I think that unit test would not block the build. Btw all non criterion tests compile file. @czanik could you please check if I am right here, and you can build sles11 using this source?

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

gaborznagy commented:
@definit3:
I have checked your code, looks good to me.

* Can you reduce the number of commits in this PR by squashing them, please?

As @furiel mentioned we will check SLES 11 compilation.

viliampucik commented:
I can confirm the compilation issue with `g_hash_table_contains` on RHEL6:

```
./lib/secret-storage/.libs//libsecret-storage.so.0: undefined reference to `g_hash_table_contains'
collect2: ld returned 1 exit status
make[2]: *** [lib/tests/test_host_resolve] Error 1
make[2]: *** Waiting for unfinished jobs....
./lib/secret-storage/.libs//libsecret-storage.so.0: undefined reference to `g_hash_table_contains'
collect2: ld returned 1 exit status
make[2]: *** [syslog-ng/syslog-ng] Error 1
/home/user/rpmbuild/BUILD/syslog-ng-3.14.1/lib/secret-storage/.libs/libsecret-storage.so: undefined reference to `g_hash_table_contains'
collect2: ld returned 1 exit status
make[2]: *** [modules/diskq/dqtool] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

gaborznagy commented:
Hi @*viliampucik,*

Unfortunately your comment is not posted on github.

Can you state please which version of RHEL6 and glib2 you have?
As far as I know there should be a glib2 2.28.

Gabor


On Tue, Mar 6, 2018 at 8:39 AM, Viliam Pucik <notifications@github.com>
wrote:

> *@viliampucik* commented on this pull request.
> ------------------------------
>
> In lib/secret-storage/secret-storage.c
> <https://github.com/balabit/syslog-ng/pull/1897#discussion_r172431138>:
>
> >    SecretStorage *secret_storage;
> -  if (!g_hash_table_contains(secret_manager, key))
> +  if (!g_hash_table_lookup_extended(secret_manager, key, ignore_orig_key, ignore_value))
>
> There is just one pitfall with RHEL6 which glib2 is missing
> g_hash_table_lookup. Any way to solve it?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/1897#discussion_r172431138>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYbP5S9DOGOBniVBXlgVWjmZZOzTm8Duks5tbj0XgaJpZM4SVOoH>
> .
>

gaborznagy commented:
Hi @viliampucik!

Unfortunately your comment is not posted on github, I repost it here:
>I can confirm the compilation issue with g_hash_table_contains on RHEL6:
./lib/secret-storage/.libs//libsecret-storage.so.0: undefined reference to `g_hash_table_contains'
collect2: ld returned 1 exit status
make[2]: *** [lib/tests/test_host_resolve] Error 1
make[2]: *** Waiting for unfinished jobs....
./lib/secret-storage/.libs//libsecret-storage.so.0: undefined reference to `g_hash_table_contains'
collect2: ld returned 1 exit status
make[2]: *** [syslog-ng/syslog-ng] Error 1
/home/user/rpmbuild/BUILD/syslog-ng-3.14.1/lib/secret-storage/.libs/libsecret-storage.so: undefined reference to `g_hash_table_contains'
collect2: ld returned 1 exit status
make[2]: *** [modules/diskq/dqtool] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

> In lib/secret-storage/secret-storage.c:
>   SecretStorage *secret_storage;
> -  if (!g_hash_table_contains(secret_manager, key))
> +  if (!g_hash_table_lookup_extended(secret_manager, key, ignore_orig_key, ignore_value))
> There is just one pitfall with RHEL6 which glib2 is missing g_hash_table_lookup. Any way to solve it?


Can you state please which version of RHEL6 and glib2 you have?
As far as I know there should be glib2 2.28 and g_hash_table_lookup was existing in 2.26 too.
There should not be a problem on RHEL6.

Gabor

viliampucik commented:
Hi @gaborznagy, I have just replaced `g_hash_table_contains` with `g_hash_table_lookup` and I was able to successfully build syslog-ng 3.1.4.1 from the official source code on RHEL6 (glib2-2.28.8-9.el6.x86_64).

gaborznagy commented:
@viliampucik: Thanks for the clarification and I'm glad there is no compilation error on RHEL6 for you.

You will not have to replace `g_hash_table_contains` with `g_hash_table_lookup` after this PR is merged.

czanik commented:
@gaborznagy @viliampucik : Took a lot more effort than planned as I also had to find a JSON-C version which still compiles on SLES 11, but finally I managed to compile syslog-ng on SLES 11 using git head + this PR.

viliampucik commented:
Many thanks @czanik, appreciated! :)

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

kira-syslogng commented:
Build SUCCESS, the tests were executed on test branch: master and test suite: functions
